src/cpp/shared/town/town.cpp

```diff
@@ -1343,4 +1343,73 @@ int __fastcall CanBuild(town *twn, int building) {
     return 0;
   }
   return 1;
+}
+
+void town::SetFaction(FACTION faction) {
+  // Shift of town graphics in relation to current graphics
+  // Setting offset in such a way that makes the town have frames from the first faction town/castle
+  int oldFactionID = this->factionID;
+  int imgOffset = -oldFactionID;
+  if(this->factionID == FACTION_CYBORG)
+    imgOffset = -6;
+  
+  // Meanwhile applying some town changes
+  this->field_55 = 10;
+  this->factionID = faction;
+
+  // Destroying buildings that don't exist in new faction
+  for(int bldIdx = BUILDING_MAX - 1; bldIdx >= 0; bldIdx--) {
+    unsigned long eligibleBuildMask = gTownEligibleBuildMask[faction];
+    int bldMask = (1 << bldIdx);
+
+    // Skip if building is a castle, not built or can be built
+    if(bldIdx == BUILDING_CASTLE || !(bldMask & this->buildingsBuiltFlags) || (bldMask & eligibleBuildMask))
+      continue;
+
+    // Remove building that can't be built
+    this->buildingsBuiltFlags &= ~(bldMask);
+
+    // Downgrade buildings if upgraded ones don't exist in new faction
+    int dwellingIdx = bldIdx - 19;
+    int numCreatures = this->numCreaturesInDwelling[dwellingIdx];
+    if(bldIdx == BUILDING_UPGRADE_5B) {
+      this->numCreaturesInDwelling[dwellingIdx] = 0;
```

Comment:
So Cyclops an change into Black Dragons, but Black Dragons change into nothing?

Reply:
No. Cyclops will change to Green Dragons. Black Dragons will change to Cyclops
The algorithm will never upgrade creatures, only downgrade